lib/internal/test_runner/utils.js

```diff
@@ -168,7 +169,10 @@ async function setupTestReporters(rootTest) {
   const reportersMap = await getReportersMap(reporters, destinations, rootTest);
   for (let i = 0; i < reportersMap.length; i++) {
     const { reporter, destination } = reportersMap[i];
-    compose(rootTest.reporter, reporter).pipe(destination);
+    compose(rootTest.reporter, reporter).on('error', (err) => {
```

Comment:
I'm not sure that this handles all errors. Can you add a test with an error thrown from a `setImmediate()` to verify.

Reply:
@cjihrig Without an architectural change (e.g running reporters from their own worker) I'm not sure how we can catch those kind of errors?

Maybe with an uncaught exception handler + checking the stack?

Reply:
In other words, we catch them in the same way we catch error from tests.

Reply:
Yes async hooks + process exit handler would work and unfortunately is probably justified in this case :D 

I would prefer we do AsyncLocalStorage + process exit handler (with the same trick) as it's getting reimplemented without async_hooks so it's better not to pick up an explicit dependency.

+1 for that.